torch/_functorch/aot_autograd.py

```diff
@@ -706,9 +834,52 @@ def is_fun(t):
 
     return isinstance(t, FunctionalTensor)
 
+# t here is either
+# (1) A FunctionalTensor(_to_functional_tensor(FakeTensor))
+# (2) A traceable tensor subclass that holds a FunctionalTensor
 def has_metadata_mutation(t):
-    assert isinstance(t, FunctionalTensor)
-    return torch._functionalize_has_metadata_mutation(t.elem)
+    if is_traceable_wrapper_subclass(t):
+        attrs, _ = t.__tensor_flatten__()
+        # A tensor subclass was updated if any of its inner elements were updated
+        return any(has_metadata_mutation(getattr(t, attr)) for attr in attrs)
+    else:
+        assert isinstance(t, FunctionalTensor)
+        return torch._functionalize_has_metadata_mutation(t.elem)
+
+# new_arg and arg here are either:
+# (1) both a FakeTensor
+# (2) both a traceable tensor subclass that holds a FakeTensor
+# Pre-condition: the two args are the "old" and "new" inputs from running functionalization.
+# When we run functionalization and wrap our inputs into FunctionalTensors,
+# we can detect whether or not an input was mutated by checking to see if the inner tensor has changed
+def was_updated(arg, new_arg):
+    if is_traceable_wrapper_subclass(arg):
```

Comment:
The point here is that we always reconstruct the outer wrapper, so you can't do an identity check on the outside, right?

Reply:
yep exactly!

Reply:
yep, although the reason isn't obvious - i'll add a note. It's because `from_fun(to_fun(x)) is x` returns False when x is a tensor subclass. If I have a `TwoTensor(FakeTensor, FakeTensor)`, then when I functionalize I have to construct an entirely new subclass, from the call to `__tensor_unflatten__`: `TwoTensor(FunctionalTensor, FunctionalTensor)`.